src/GafferUI/View.cpp

```diff
@@ -268,16 +268,36 @@ void View::toolPlugSet( Gaffer::Plug *plug ) const
 		return;
 	}
 
-	if( !tool->activePlug()->getValue() || !exclusive( tool ) )
+	if( !exclusive( tool ) )
 	{
 		return;
 	}
 
+	if( tool->activePlug()->getValue() )
+	{
+		for( auto &t : Tool::Range( *tools() ) )
+		{
+			if( t != tool && exclusive( t.get() ) )
+			{
+				t->activePlug()->setValue( false );
+			}
+		}
+	}
+
+	bool exclusivePlugActive = false;
 	for( auto &t : Tool::Range( *tools() ) )
 	{
-		if( t != tool && exclusive( t.get() ) )
+		exclusivePlugActive |= ( exclusive( t.get() ) && t->activePlug()->getValue() );
+	}
+	if( !exclusivePlugActive )
```

Comment:
I think this can just be an `else` block on the `if( tool->activePlug()->getValue() )` conditional can't it? We already know that `tool` is exclusive, so if we know that it has been turned off, then that's all we need isn't it?

Reply:
That does make sense - thanks for the clear explanation.

We can change the const-ness of `toolPlugSet()` if we want though - because it's private nobody can be depending on it. What we can't do is add a member, which is unfortunately something we'd want if we were to store a `Signals::Connection` to block, or store a flag to say we're currently calling `setValue()` ourselves.

I does seem like everything would be more predictable if we avoided re-entrancy in `toolPlugSet()` - I'm sure it took some figuring out to determine why the obvious approach didn't work. Since this is all happening on the UI thread, and it's reasonable to assume that it's all local to one View (mucking with the active plugs won't affect another View's plugs), maybe we can just use a static `bool g_ignoreToolPlugSet` that we use to avoid reentrancy? Or we could have that as a member variable and move this to `1.4`. Any preference? Maybe you'd rather stick to what you have since it's tested and working?

Reply:
I think for clarity I'd prefer to store a `Signals::Connection` so we can block signals when we're doing our default fallback tool selection. Thinking about myself or someone else coming back to this, that sounds like the clearest explanation of what's happening.

It doesn't seem like it's too much of a priority, so pushing it into `1.4` wouldn't be too big of a deal?

If I'm wrong about the priority, the static `bool g_ignoreToolPlugSet` should be fine, I just have a nag in the back of my mind that makes me wary of storing local state in a global variable.

So I'm happy to update an retest, and which solution we go with seems dictated by the priority. Any preference from you on including it in `1.4` vs `1.3`?

Reply:
> I think for clarity I'd prefer to store a Signals::Connection so we can block signals when we're doing our default fallback tool selection.

One wrinkle with this would be that we'd need a connection for each tool, which would be a bit more faffy than I made it sound at first. But it's definitely doable.

> I just have a nag in the back of my mind that makes me wary of storing local state in a global variable.

For good reason too! But I can't see a way it would fail in this case.

> Any preference from you on including it in 1.4 vs 1.3?

No huge preference - I agree this isn't a high priority item, so if you'd just like to do the clean version and put it in 1.4 that's fine with me.

Reply:
Commit https://github.com/GafferHQ/gaffer/pull/5518/commits/34a7926dd0b395ac854135772f5ec5a783c92493 changes the method for activating the default tool to use a `Signal::BlockedConnection` instead of looking to see if all tools were deactivated.

It feels like a pretty good solution...